cloudwatch_logs_common/src/log_publisher.cpp

```diff
@@ -108,102 +107,153 @@ Aws::CloudWatchLogs::ROSCloudWatchLogsErrors LogPublisher::StopPublisherThread()
+    AWS_LOG_DEBUG(__func__, "Successfully created log stream.");
+  } else if (CW_LOGS_LOG_STREAM_ALREADY_EXISTS == create_log_stream_status) {
+    this->run_state_ = Aws::CloudWatchLogs::LOG_PUBLISHER_RUN_INIT_TOKEN;
+    AWS_LOG_DEBUG(__func__, "Log stream already exists, quit attempting to create it.");
+  } else {
+    AWS_LOGSTREAM_ERROR(__func__, "Failed to create log stream, status: " 
+                        << create_log_stream_status << ". Retrying ...");
+  }
+}
 
-    this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_, next_token);
-
-    auto shared_logs_obj = this->shared_logs_.load(std::memory_order_acquire);
-    if (nullptr != shared_logs_obj) {
-      std::list<Aws::CloudWatchLogs::Model::InputLogEvent> * logs;
-      publisher_status = shared_logs_obj->getDataAndLock(logs);
-      if (CW_LOGS_SUCCEEDED == publisher_status) {
-        if (!logs->empty()) {
-          Aws::CloudWatchLogs::ROSCloudWatchLogsErrors send_logs_status = CW_LOGS_FAILED;
-          int tries = kMaxRetries;
-
-          while (CW_LOGS_SUCCEEDED != send_logs_status && tries > 0) {
-            send_logs_status = this->cloudwatch_facade_->SendLogsToCloudWatch(
-              next_token, this->log_group_, this->log_stream_, logs);
-            if (CW_LOGS_SUCCEEDED != send_logs_status) {
-              AWS_LOG_WARN(__func__, "Unable to send logs to CloudWatch, retrying ...");
-              Aws::CloudWatchLogs::ROSCloudWatchLogsErrors get_token_status =
-                this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_,
-                                                            next_token);
-              if (CW_LOGS_SUCCEEDED != get_token_status) {
-                AWS_LOG_WARN(__func__,
-                             "Unable to obtain the sequence token to use, quit attempting to send "
-                             "logs to CloudWatch");
-                break;
-              }
-            }
-            tries--;
-          }
+/**
+ * Fetches the token to use for writing logs to a stream. 
+ */
+void LogPublisher::InitToken(Aws::String & next_token)
+{
+  Aws::CloudWatchLogs::ROSCloudWatchLogsErrors get_token_status =
+    this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_,
+                                                next_token);
+  if (CW_LOGS_SUCCEEDED == get_token_status) {
+    this->run_state_ = Aws::CloudWatchLogs::LOG_PUBLISHER_RUN_SEND_LOGS;
+  } else {
+    AWS_LOGSTREAM_ERROR(__func__, "Unable to obtain the sequence token to use, status: " 
+                        << get_token_status << ". Retrying ...");
+  }
+}
+
+/**
+ * Checks if there are logs ready to be sent out. If so then it attempts to send them
+ * to CloudWatch.
+ */
+void LogPublisher::SendLogs(Aws::String & next_token)
+{
+  Aws::CloudWatchLogs::ROSCloudWatchLogsErrors publisher_status = CW_LOGS_SUCCEEDED;
+  auto shared_logs_obj = this->shared_logs_.load(std::memory_order_acquire);
```

Comment:
The avoidance of a lock here was an optimization around the shared object that the producer is writing into and the consumer is reading from. It may have been somewhat premature, but as it was existing I wouldn't want to change it for this PR.